src/cunumeric/sort/sort.cu

```diff
@@ -95,277 +137,1514 @@ struct SortPiece {
+    CHECK_CUDA(cudaMemcpy2DAsync(size_recv_total.ptr(0),
+                                 1 * sizeof(size_t),
+                                 size_recv.ptr(num_segments_l),
+                                 num_segments_l_aligned * sizeof(size_t),
+                                 sizeof(int64_t),
+                                 num_sort_ranks,
+                                 cudaMemcpyDeviceToHost,
+                                 stream));
+
+    // need to sync as we share values in between host/device
+    CHECK_CUDA(cudaStreamSynchronize(stream));
   }
 
   // copy values into aligned send buffer
-  auto val_send_buf = local_sorted.values;
-  if (buf_size_send_vals_total > volume) {
-    val_send_buf = create_buffer<VAL>(buf_size_send_vals_total, Memory::GPU_FB_MEM);
-    size_t pos   = 0;
-    for (size_t r = 0; r < num_ranks; ++r) {
-      CHECK_CUDA(cudaMemcpyAsync(val_send_buf.ptr(aligned_pos_vals_send[r]),
-                                 local_sorted.values.ptr(pos),
-                                 sizeof(VAL) * size_send[r],
-                                 cudaMemcpyDeviceToDevice,
-                                 stream));
-      pos += size_send[r];
+  std::vector<Buffer<VAL>> val_send_buffers(num_sort_ranks);
+  std::vector<Buffer<int64_t>> idc_send_buffers(num_sort_ranks);
+  {
+    for (size_t r = 0; r < num_sort_ranks; r++) {
+      val_send_buffers[r] = create_buffer<VAL>(size_send_total[r], Memory::GPU_FB_MEM);
+      if (argsort) {
+        idc_send_buffers[r] = create_buffer<int64_t>(size_send_total[r], Memory::GPU_FB_MEM);
+      }
     }
-    local_sorted.values.destroy();
-  }
 
-  // copy indices into aligned send buffer
-  auto idc_send_buf = local_sorted.indices;
-  if (argsort && buf_size_send_idcs_total > volume) {
-    idc_send_buf = create_buffer<int64_t>(buf_size_send_idcs_total, Memory::GPU_FB_MEM);
-    size_t pos   = 0;
-    for (size_t r = 0; r < num_ranks; ++r) {
-      CHECK_CUDA(cudaMemcpyAsync(idc_send_buf.ptr(aligned_pos_idcs_send[r]),
-                                 local_sorted.indices.ptr(pos),
-                                 sizeof(int64_t) * size_send[r],
-                                 cudaMemcpyDeviceToDevice,
-                                 stream));
-      pos += size_send[r];
+    {
+      Buffer<VAL*> val_send_buffers_ptr = create_buffer<VAL*>(num_sort_ranks, Memory::Z_COPY_MEM);
+      for (size_t r = 0; r < num_sort_ranks; r++) {
+        val_send_buffers_ptr[r] = val_send_buffers[r].ptr(0);
+      }
+
+      // is either 1, 2, 4, ... , THREADS_PER_BLOCK/2
+      int segments_per_threadblock =
+        compute_segments_per_threadblock(segment_size_l, num_segments_l);
+      dim3 block_shape =
+        dim3(THREADS_PER_BLOCK / segments_per_threadblock, segments_per_threadblock);
+      const size_t num_blocks_y = (num_segments_l + block_shape.y - 1) / block_shape.y;
+      const size_t num_blocks_x = ((segment_size_l + 31) / 32 + block_shape.x - 1) / block_shape.x;
```

Comment:
can you leave a comment on the purpose of this heuristic?

Reply:
The 32 does not refer to the warp size. It is the maximum number of elements every thread is responsible for. In case there is more than (32*threads) elements to be handled in a segment the amount of blocks assigned per segment will be increased.

This should only trigger in case we have large segments (extreme case is 1D). Here I want to have multiple blocks working on the data and not just one to saturate memory bandwidth.

Reply:
in case where the segment is small, is this still better than doing `(segment_size_l + block_shape.x - 1) / block_shape.x` (i.e., each thread getting assigned to only one element)? I think a magic number like 32 needs an explanation for anyone, including yourself, who would need to touch this code later.

Reply:
In case the segments are small (and we have many of them) the threadblocks should already cover multiple segments which leaves only a fraction of threads per segment. The magic-32 only comes into play once we have large rows and decide that we need multiple blocks per row.